### PR TITLE
Clear method store / query cache confusion

### LINKED ISSUE #18150: Method store and query cache confusion in our code
Deep in the core of libcrypto, we have library context wide method stores for all methods constructed from provided implementation.
Associated with each method store, we also have a query cache, which caches the result from a combined \< algorithm, propquery \> tuple of individual method fetches.

The method stores and their contents are designed to be fairly permanent, i.e. have a lifespan similar to the providers that provide the implementations.  This lifespan is in turn within the lifespan of the library context in which the providers are loaded and the method stores created.
The query cache, on the other hand, can have a shorter lifespan...  it's basically flushed as soon as the correctness of the cache can no longer be guaranteed (for example, when loading another provider, which may give different fetch results for the same fetch input).

It seems, however, that the method store and the query cache have been confused with each other in places.  The result is that in some places, where the intention is clearly to just flush the cache, the corresponding method store is emptied!  (and furthermore, the records of having constructed methods is cleared, for *all* providers in a library context)  Under some circumstances, this leads to methods being reconstructed over and over.

### PR DESCRIPTION
**Pull request by levitte**
Stop emptying the method stores or clearing the corresponding provider
synchronisation records (operation bits) when flushing the query cache.
Rename some internal things to better express what they are for.

This also refactors the core method construction loop so the provider
synchronisation records (operation bits) are correctky tested.

Fixes #18150

### CODE REVIEW

**Comment by levitte** [comment]
Ah, I forgot to update the FIPS checksums (and also to actually test for myself with FIPS enabled).

> I assume the intention is to backport this to 3.0 since this is a serious bug there?

Yes.  Labels updated.

**Comment by levitte** [comment]
So from some FIPS related failures, it looks like there is actually an expectation that a method store gets cleared under some pretty special circumstances, all revolving around OSSL_SELF_TEST

**Comment by levitte** [comment]
Making this a draft for some time.  It appears that there's one bit that actually needs to be able to drop the methods of a provider:

> **OSSL_PROVIDER_self_test**() is used to run a provider's self tests on
> demand.  If the self tests fail then the provider will fail to provide
> any further services and algorithms. **OSSL_SELF_TEST_set_callback**(3) may
> be called beforehand in order to display diagnostics for the running
> self tests.

Although, I'm not quite sure if clearing the method store is necessary or optimal, or from another point of view, if it's enough.

**Comment by t8m** [comment]
> Although, I'm not quite sure if clearing the method store is necessary or optimal, or from another point of view, if it's enough.

IMO it makes sense. If after that the failing provider just won't respond to any query operations anymore. Only methods for the particular provider should be cleared though.

**Comment by levitte** [comment]
> > Although, I'm not quite sure if clearing the method store is necessary or optimal, or from another point of view, if it's enough.
> 
> IMO it makes sense. If after that the failing provider just won't respond to any query operations anymore. Only methods for the particular provider should be cleared though.

Yeah...  I had a chat with @mattcaswell and we figured out together what's actually going on and why.  Dropping the methods is the correct response for this...  but should be done selectively, i.e. only those associated with the provider for which the self test failed.

**Comment by levitte** [comment]
I've added a fixup commit for a few details, and another commit that adds clearly defined functionality to remove all methods that belong to a specified provider.  The latter is used to restore `ossl_provider_self_test()`'s intended functionality.

**Comment by levitte** [comment]
Tests are passing on my laptop, so I'm making this PR ready to review again

File: crypto/provider_core.c [hunk 1]
```diff
@@ -1152,7 +1152,25 @@ static int provider_flush_store_cache(const OSSL_PROVIDER *prov)
     CRYPTO_THREAD_unlock(store->lock);
 
     if (!freeing)
-        return evp_method_store_flush(prov->libctx);
+        return evp_method_store_cache_flush(prov->libctx);
```
Comment (levitte) [inline crypto/provider_core.c:1155]: One thing I found infuriating is that we do this for EVP stuff only, but not OSSL_STORE, OSSL_ENCODER or OSSL_DECODER.  I'm *not* addition the other implementation types in this PR, but to be complete, we really should, and also probably make a mechanism that helps us avoid this sort of hard coding, as it's just too easy to forget the gazillion places to touch as soon as a new (non_EVP) implementation type comes along

File: crypto/provider_core.c [hunk 2]
```diff
@@ -1156,6 +1156,24 @@ static int provider_flush_store_cache(const OSSL_PROVIDER *prov)
     return 1;
 }
 
+static int provider_empty_store(const OSSL_PROVIDER *prov)
```
Comment (t8m) [inline crypto/provider_core.c:1159]: The name is a little bit confusing to me. Maybe `provider_drop_from_store` ?

**Comment by t8m** [comment]
CI is relevant.

**Comment by levitte** [comment]
Yikes, memory stuff.  I'll deal with that on Monday, I'm pretty finished for today

File: crypto/provider_core.c (hunk 2 above)
Comment (levitte) [reply crypto/provider_core.c]: Name change done

**Comment by mattcaswell** [comment]
Looks ok to me, subject to the memory stuff in CIs being sorted.

**Comment by mattcaswell** [comment]
It would be good to know what impact this has on the decoder performance.

**Comment by levitte** [comment]
> It would be good to know what impact this has on the decoder performance.

The only measurement I personally have is a huge reduction of method constructions when calling `OSSL_DECODER_CTX_new_for_pkey()` multiple times with the same input, i.e. the methods are only constructed once instead of every time.  @paulidale has a bigger thing that I hope he'll try this PR on.

**Comment by levitte** [comment]
Only the address_ub_sanitizer job that's still failing.  That's some sort of progress at least.
Only having "Indirect leaks" indicates that there's some sort of mutual reference between two allocated objects, but no root object pointing at them...  that's most likely some more cleanup that doesn't do its job quite right

**Comment by levitte** [comment]
@t8m, could the memory issue I encounter here possibly be related to #18139 (fixed in #18141)?

**Comment by t8m** [comment]
@levitte I wouldn't think so. As there just the space allocated by LHASH is being grown and not those objects pointed to it.

**Comment by levitte** [comment]
Something I noted today is that for `15-test_rsa.t`, the only runs with ub sanitizer failures are those involving PVK...  that makes me think that the issue isn't in the core stuff I touch here, but that there's a leak somewhere else that has laid dormant 'til now.

**Comment by levitte** [comment]
Looking at `05-test_pbe.t` and `test/pbetest.c`, one can note that there are a lot of leaks involving `test_pkcs5_pbe_rc4_md5`, and none involving `test_pkcs5_pbe_des_sha1`.

This, together with the leaks revolving around PVK in the `15-test_rsa.t` runs, tells me to look more closely at the legacy provider

File: crypto/property/property.c [hunk 3]
```diff
@@ -360,6 +378,56 @@ int ossl_method_store_remove(OSSL_METHOD_STORE *store, int nid,
...
+    const OSSL_PROVIDER *prov;
+};
+
+static void
+alg_cleanup_by_provider(ossl_uintmax_t idx, ALGORITHM *alg, void *arg)
+{
+    struct alg_cleanup_by_provider_data_st *data = arg;
+    int i, count;
+
+
```
Comment (paulidale) [inline crypto/property/property.c:392]: Spurious blank line.

Comment (levitte) [reply crypto/property/property.c]: Fixed

**Comment by levitte** [comment]
I've had a couple of interactions with @mattcaswell to try and understand what's going on with the child libctx / provider stuff.  We had a discussion on possible ideas yesterday, and he came up with one that seems to be the actual answer, see 82ba5253257fcb73c45a46134f4aa72c14af5cdc

**Comment by levitte** [comment]
(famous last words: it works in my sanitizer tests :wink:)

**Comment by levitte** [comment]
> Given the amount of discussion we had to come up with this, this is probably worth a comment.

...  such as in CHANGES.md, is that what you mean?

**Comment by levitte** [comment]
> Given the amount of discussion we had to come up with this, this is probably worth a comment.

Comment added in `crypto/provider_child.c`

**Comment by levitte** [comment]
Done a clean rebase on a fresh master

**Comment by levitte** [comment]
> There is an interesting failure in the CI.

Darnit...

**Comment by levitte** [comment]
> I suppose it is due to OSSL_PROVIDER_unload not fully enforcing the unload still. There are some testcases in the test that load the default provider into the default context and later unloading it. With this PR this unload does not have the full effect it seems.

Not sure what you mean with "not fully enforcing".  Is that a command against activation count or against refcount?

**Comment by t8m** [comment]
I mean the fetch apparently is able to fetch things from the default provider although we called unload. Not sure what would keep holding the reference here. It does not look like it should be held.

**Comment by levitte** [comment]
The fault with `30-test_evp_extra.t` remains.

Initial observation: this only happens when `enable-fips` is configured.

**Comment by t8m** [comment]
> The fault with `30-test_evp_extra.t` remains.
> 
> Initial observation: this only happens when `enable-fips` is configured.
> 
> UPDATE: But apparently not just that. I tried with this configuration on my machine, and it still passes:
> 
> ```
> $ ./configdata.pm -c
> 
> Command line (with current working directory = .):
> 
>     /usr/bin/perl ../master/Configure --debug --strict-warnings enable-asan enable-ubsan enable-fips
> ```

Have you used `OPENSSL_TEST_RAND_ORDER=1651232502`

**Comment by levitte** [comment]
Er, I can't even reproduce the `address_ub_sanitizer` failure.

``` console
$ (cd ../master; git rev-parse HEAD)
64bc496325a5d7f573921aa68317d0f0a4e0390c
$ make clean; ../master/Configure --debug enable-asan enable-ubsan enable-rc5 enable-md2 enable-ec_nistp_64_gcc_128 enable-fips -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
$ make -j8
...
$ make test
make depend && make _tests
make[1]: Entering directory '/home/levitte/gitwrk/openssl.net/official/_build'
make[1]: Leaving directory '/home/levitte/gitwrk/openssl.net/official/_build'
make[1]: Entering directory '/home/levitte/gitwrk/openssl.net/official/_build'
( SRCTOP=../master \
  BLDTOP=. \
  PERL="/usr/bin/perl" \
  FIPSKEY="f4556650ac31d35461610bac4ed81b1a181b2d8a43ea2854cbae22ca74560813" \
  EXE_EXT= \
  /usr/bin/perl ../master/test/run_tests.pl  )
00-prep_fipsmodule_cnf.t .. ok
All tests successful.
Files=1, Tests=1,  1 wallclock secs ( 0.01 usr  0.00 sys +  0.25 cusr  0.05 csys =  0.31 CPU)
Result: PASS
01-test_abort.t .................... ok
01-test_fipsmodule_cnf.t ........... ok
01-test_sanity.t ................... ok
01-test_symbol_presence.t .......... ok
01-test_test.t ..................... ok
02-test_errstr.t ................... ok
02-test_internal_context.t ......... ok
02-test_internal_ctype.t ........... ok
02-test_internal_keymgmt.t ......... ok
02-test_internal_provider.t ........ ok
02-test_lhash.t .................... ok
02-test_localetest.t ............... ok
02-test_ordinals.t ................. ok
02-test_sparse_array.t ............. ok
02-test_stack.t .................... ok
03-test_exdata.t ................... ok
03-test_fipsinstall.t .............. ok
03-test_internal_asn1.t ............ ok
03-test_internal_asn1_dsa.t ........ ok
03-test_internal_bn.t .............. ok
03-test_internal_chacha.t .......... ok
03-test_internal_curve448.t ........ ok
03-test_internal_ec.t .............. ok
03-test_internal_ffc.t ............. ok
03-test_internal_mdc2.t ............ ok
03-test_internal_modes.t ........... ok
03-test_internal_namemap.t ......... ok
03-test_internal_poly1305.t ........ ok
03-test_internal_rsa_sp800_56b.t ... ok
03-test_internal_siphash.t ......... ok
03-test_internal_sm2.t ............. ok
03-test_internal_sm3.t ............. ok
03-test_internal_sm4.t ............. ok
03-test_internal_ssl_cert_table.t .. ok
03-test_internal_x509.t ............ ok
03-test_params_api.t ............... ok
03-test_property.t ................. ok
03-test_ui.t ....................... ok
04-test_asn1_decode.t .............. ok
04-test_asn1_encode.t .............. ok
04-test_asn1_string_table.t ........ ok
04-test_bio_callback.t ............. ok
04-test_bio_core.t ................. ok
04-test_bio_tfo.t .................. skipped: This test requires enable-tfo
04-test_bioprint.t ................. ok
04-test_conf.t ..................... ok
04-test_encoder_decoder.t .......... ok
04-test_encoder_decoder_legacy.t ... ok
04-test_err.t ...................... ok
04-test_hexstring.t ................ ok
04-test_param_build.t .............. ok
04-test_params.t ................... ok
04-test_params_conversion.t ........ ok
04-test_pem_read_depr.t ............ ok
04-test_pem_reading.t .............. ok
04-test_provfetch.t ................ ok
04-test_provider.t ................. ok
04-test_provider_fallback.t ........ ok
04-test_provider_pkey.t ............ ok
04-test_upcalls.t .................. ok
05-test_bf.t ....................... ok
05-test_cast.t ..................... ok
05-test_cmac.t ..................... ok
05-test_des.t ...................... ok
05-test_hmac.t ..................... ok
05-test_idea.t ..................... ok
05-test_pbe.t ...................... ok
05-test_rand.t ..................... ok
05-test_rc2.t ...................... ok
05-test_rc4.t ...................... ok
05-test_rc5.t ...................... ok
06-test_algorithmid.t .............. ok
06-test_rdcpu_sanity.t ............. ok
10-test_bn.t ....................... ok
10-test_exp.t ...................... ok
15-test_dh.t ....................... ok
15-test_dsa.t ...................... ok
15-test_dsaparam.t ................. ok
15-test_ec.t ....................... ok
15-test_ecdsa.t .................... ok
15-test_ecparam.t .................. ok
15-test_gendh.t .................... ok
15-test_gendhparam.t ............... ok
15-test_gendsa.t ................... ok
15-test_genec.t .................... ok
15-test_genrsa.t ................... ok
15-test_mp_rsa.t ................... ok
15-test_out_option.t ............... ok
15-test_rsa.t ...................... ok
15-test_rsaoaep.t .................. ok
15-test_rsapss.t ................... ok
15-test_sha.t ...................... ok
20-test_app.t ...................... ok
20-test_cli_fips.t ................. ok
20-test_dgst.t ..................... ok
20-test_dhparam.t .................. ok
20-test_dhparam_check.t ............ ok
20-test_enc.t ...................... ok
20-test_enc_more.t ................. ok
20-test_kdf.t ...................... ok
20-test_mac.t ...................... ok
20-test_passwd.t ................... ok
20-test_pkeyutl.t .................. ok
20-test_rand_config.t .............. ok
20-test_spkac.t .................... ok
25-test_crl.t ...................... ok
25-test_d2i.t ...................... ok
25-test_eai_data.t ................. ok
25-test_pkcs7.t .................... ok
25-test_req.t ...................... ok
25-test_rusext.t ................... ok
25-test_sid.t ...................... ok
25-test_verify.t ................... ok
25-test_verify_store.t ............. ok
25-test_x509.t ..................... ok
30-test_acvp.t ..................... ok
30-test_aesgcm.t ................... ok
30-test_afalg.t .................... ok
30-test_defltfips.t ................ ok
30-test_engine.t ................... ok
30-test_evp.t ...................... ok
30-test_evp_extra.t ................ ok
30-test_evp_fetch_prov.t ........... ok
30-test_evp_kdf.t .................. ok
30-test_evp_libctx.t ............... ok
30-test_evp_pkey_dparam.t .......... ok
30-test_evp_pkey_provided.t ........ ok
30-test_pbelu.t .................... ok
30-test_pkey_meth.t ................ ok
30-test_pkey_meth_kdf.t ............ ok
30-test_prov_config.t .............. ok
30-test_provider_status.t .......... ok
40-test_rehash.t ................... ok
60-test_x509_check_cert_pkey.t ..... ok
60-test_x509_dup_cert.t ............ ok
60-test_x509_store.t ............... ok
60-test_x509_time.t ................ ok
61-test_bio_prefix.t ............... ok
61-test_bio_readbuffer.t ........... ok
65-test_cmp_asn.t .................. ok
65-test_cmp_client.t ............... ok
65-test_cmp_ctx.t .................. ok
65-test_cmp_hdr.t .................. ok
65-test_cmp_msg.t .................. ok
65-test_cmp_protect.t .............. ok
65-test_cmp_server.t ............... ok
65-test_cmp_status.t ............... ok
65-test_cmp_vfy.t .................. ok
66-test_ossl_store.t ............... ok
70-test_asyncio.t .................. ok
70-test_bad_dtls.t ................. ok
70-test_clienthello.t .............. ok
70-test_comp.t ..................... ok
70-test_key_share.t ................ ok
70-test_packet.t ................... ok
70-test_recordlen.t ................ ok
70-test_renegotiation.t ............ ok
70-test_servername.t ............... ok
70-test_sslcbcpadding.t ............ ok
70-test_sslcertstatus.t ............ ok
70-test_sslextension.t ............. ok
70-test_sslmessages.t .............. ok
70-test_sslrecords.t ............... ok
70-test_sslsessiontick.t ........... ok
70-test_sslsigalgs.t ............... ok
70-test_sslsignature.t ............. ok
70-test_sslskewith0p.t ............. ok
70-test_sslversions.t .............. ok
70-test_sslvertol.t ................ ok
70-test_tls13alerts.t .............. ok
70-test_tls13cookie.t .............. ok
70-test_tls13downgrade.t ........... ok
70-test_tls13hrr.t ................. ok
70-test_tls13kexmodes.t ............ ok
70-test_tls13messages.t ............ ok
70-test_tls13psk.t ................. ok
70-test_tlsextms.t ................. ok
70-test_verify_extra.t ............. ok
70-test_wpacket.t .................. ok
71-test_ssl_ctx.t .................. ok
79-test_http.t ..................... ok
80-test_ca.t ....................... ok
80-test_ca_internals.t ............. ok
80-test_cipherbytes.t .............. ok
80-test_cipherlist.t ............... ok
80-test_ciphername.t ............... ok
80-test_cmp_http.t ................. skipped: These tests are not supported in a fuzz build
80-test_cms.t ...................... ok
80-test_cmsapi.t ................... ok
80-test_ct.t ....................... ok
80-test_dane.t ..................... ok
80-test_dtls.t ..................... ok
80-test_dtls_mtu.t ................. ok
80-test_dtlsv1listen.t ............. ok
80-test_ocsp.t ..................... ok
80-test_pkcs12.t ................... ok
80-test_ssl_new.t .................. ok
80-test_ssl_old.t .................. ok
80-test_ssl_test_ctx.t ............. ok
80-test_sslcorrupt.t ............... ok
80-test_tsa.t ...................... ok
80-test_x509aux.t .................. ok
81-test_cmp_cli.t .................. skipped: These tests are not supported in a fuzz build
82-test_tfo_cli.t .................. ok
90-test_asn1_time.t ................ ok
90-test_async.t .................... ok
90-test_bio_enc.t .................. ok
90-test_bio_memleak.t .............. ok
90-test_constant_time.t ............ ok
90-test_fatalerr.t ................. ok
90-test_fipsload.t ................. skipped: Test is disabled in an address sanitizer build
90-test_gmdiff.t ................... ok
90-test_ige.t ...................... ok
90-test_includes.t ................. ok
90-test_memleak.t .................. ok
90-test_overhead.t ................. ok
90-test_secmem.t ................... ok
90-test_shlibload.t ................ skipped: Test is disabled in an address sanitizer build
90-test_srp.t ...................... ok
90-test_sslapi.t ................... ok
90-test_sslbuffers.t ............... ok
90-test_store.t .................... ok
90-test_sysdefault.t ............... ok
90-test_threads.t .................. ok
90-test_time_offset.t .............. ok
90-test_tls13ccs.t ................. ok
90-test_tls13encryption.t .......... ok
90-test_tls13secrets.t ............. ok
90-test_v3name.t ................... ok
91-test_pkey_check.t ............... ok
95-test_external_gost_engine.t ..... skipped: No external tests in this configuration
95-test_external_krb5.t ............ skipped: No external tests in this configuration
95-test_external_oqsprovider.t ..... skipped: No external tests in this configuration
95-test_external_pyca.t ............ skipped: No external tests in this configuration
95-test_external_tlsfuzzer.t ....... skipped: No external tests in this configuration
99-test_ecstress.t ................. ok
99-test_fuzz_asn1.t ................ ok
99-test_fuzz_asn1parse.t ........... ok
99-test_fuzz_bignum.t .............. ok
99-test_fuzz_bndiv.t ............... ok
99-test_fuzz_client.t .............. ok
99-test_fuzz_cmp.t ................. ok
99-test_fuzz_cms.t ................. ok
99-test_fuzz_conf.t ................ ok
99-test_fuzz_crl.t ................. ok
99-test_fuzz_ct.t .................. ok
99-test_fuzz_server.t .............. ok
99-test_fuzz_x509.t ................ ok
All tests successful.
Files=248, Tests=3503, 874 wallclock secs ( 7.66 usr  0.56 sys + 759.71 cusr 129.52 csys = 897.45 CPU)
Result: PASS
make[1]: Leaving directory '/home/levitte/gitwrk/openssl.net/official/_build'
```

Can someone else reproduce the failure?

**Comment by levitte** [comment]
> Have you used `OPENSSL_TEST_RAND_ORDER=1651232502`?

Ah, the thing I always forget.  Trying it now

**Comment by levitte** [comment]
> > Have you used `OPENSSL_TEST_RAND_ORDER=1651232502`?
> 
> Ah, the thing I always forget. Trying it now

That did reproduce the failure...  and that helps a lot, I can zero in on interactions between test cases in `test/evp_extra_test2.c`

**Comment by levitte** [comment]
With a hacked `test/evp_extra_test2.c`, I've finally determined that there's a bad interaction between `test_pkcs8key_nid_bio` and `test_alternative_default`

**Comment by levitte** [comment]
It's been hinted to me that there's something weird going on with loading and unloading the "default" provider.  That can't explain everything, though, 'cause `test_d2i_PrivateKey_ex` - which also loads/unloads the "default" provider - interacts well with `test_alternative_default`, at least in my runs

**Comment by levitte** [comment]
I've found that running `test_pkcs8key_nid_bio` before `test_alternative_default` is what causes the failure.

**Comment by levitte** [comment]
> I mean the fetch apparently is able to fetch things from the default provider although we called unload. Not sure what would keep holding the reference here. It does not look like it should be held.

I finally caught up on what you said there...  it seems I had removed a little too much, so a an deactivation didn't remove methods from the method store any more...  I think I'm finally clearing the last bug in this PR (famous last words, I know...)

**Comment by levitte** [comment]
I've done some mods, and well, it passes on my machine (also famous last words)

**Comment by levitte** [comment]
I've also run tests with `OPENSSL_TEST_RAND_ORDER=1651232502`.  Still passing

**Comment by levitte** [comment]
... aaaand CIs are happy!

Quick, someone approve before it starts failing again!  🤣

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Thinking about tests - it might be nice to have an explicit testcase that would cover the case failing with OPENSSL_TEST_RAND_ORDER=1651232502.

**Review by mattcaswell — APPROVED** [review]

**Comment by levitte** [comment]
> Thinking about tests - it might be nice to have an explicit testcase that would cover the case failing with OPENSSL_TEST_RAND_ORDER=1651232502.

Heh, I guess there will be a list of numbers after a while :wink:

**Comment by t8m** [comment]
> Heh, I guess there will be a list of numbers after a while wink

I mean doing the OSSL_PROVIDER_load(), then doing something that triggered the failure, OSSL_PROVIDER_unload(), and trying to fetch the cached/stored thing from the provider that should be cleaned up and not fetchable anymore.

**Comment by levitte** [comment]
> > Heh, I guess there will be a list of numbers after a while wink
> 
> I mean doing the OSSL_PROVIDER_load(), then doing something that triggered the failure, OSSL_PROVIDER_unload(), and trying to fetch the cached/stored thing from the provider that should be cleaned up and not fetchable anymore.

Feel free, I look forward to the PR!

**Comment by levitte** [comment]
Of course we forgot the label.  I'm counting the 24h grace period from the last approval, though (more than 24hs ago)

**Comment by levitte** [comment]
Merged

master:
20b6d85ab2b9cfa4cd29d2422d69c3e3f4db0a41 Drop ossl_provider_clear_all_operation_bits() and all uses of it
10937d5867039afbf869c8514245ed7599b61307 Refactor method construction pre- and post-condition
60640d79ca7ea0980dc09c71fe6a297b5f8588a2 Don't empty the method store when flushing the query cache
2e4d0677ea858c619a33235265dbee19520a9d35 Make it possible to remove methods by the provider that provides them
03454ba2a234197c961920f1bac37cc9f4cf3f54 Complete the cleanup of an algorithm in OSSL_METHOD_STORE
4da7663b02bf05542830e85db6f74cf90daf1f49 For child libctx / provider, don't count self-references in parent
32e3c071373280b69be02ba91fc3204495e2e1bf Add method store cache flush and method removal to non-EVP operations

3.0:
749691374ccf87418ca0e6664a9b9c831dcfe108 Drop ossl_provider_clear_all_operation_bits() and all uses of it
27c4ac706a059aab81f3d1d8cc0ba0b3693bf155 Refactor method construction pre- and post-condition
8b76db9e2618225f78c84ed2b338cd005f6ef310 Don't empty the method store when flushing the query cache
215708c53bc3d62fd29ec842d80fa929910178ee Make it possible to remove methods by the provider that provides them
106f5c231ccf7103564c8f81f169e372c444351b Complete the cleanup of an algorithm in OSSL_METHOD_STORE
a860a5807adf61f624872fa26aeb31585c17aa72 For child libctx / provider, don't count self-references in parent
37a6e9efe013f9e6a840e38beb81e44b9fee3629 Add method store cache flush and method removal to non-EVP operations

**Pull request closed without merging**